cmd/migratestats/migrate.go

```diff
@@ -0,0 +1,226 @@
+package main
+
+import (
+	"log"
+	"strconv"
+	"strings"
+	"time"
+
+	errgo "gopkg.in/errgo.v1"
+	"gopkg.in/juju/charmrepo.v3/csclient/params"
+	"gopkg.in/juju/charmstore.v5/internal/charmstore"
+	"gopkg.in/juju/charmstore.v5/internal/mongodoc"
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+const (
+	statCountersColl = "juju.stat.counters"
+	statTokensColl   = "juju.stat.tokens"
+	migrationsColl   = "migrations"
+)
+
+var keepKinds = []string{
+	"archive-download",
+	"archive-download-promulgated",
+	"archive-delete",
+	"archive-upload",
+}
+
+// squashStats combines entries in the stats counters
+// into 1-hour granularity rather than 1-minute granularity.
+func squashStats(db *mgo.Database) error {
+	seconds := int(charmstore.StatsGranularity / time.Second)
+	var r []struct{}
+	err := db.C(statCountersColl).Pipe([]bson.M{{
+		"$group": bson.M{
+			"_id": bson.M{
+				"t": bson.M{
+					"$subtract": []interface{}{
+						"$t",
+						bson.M{
+							"$mod": []interface{}{"$t", seconds},
+						},
+					},
+				},
+				"k": "$k",
+			},
+			"id": bson.M{
+				"$first": "$_id",
+			},
+			"c": bson.M{
+				"$sum": "$c",
+			},
+		},
+	}, {
+		"$project": bson.M{
+			"_id": "$id",
+			"c":   "$c",
+			"k":   "$_id.k",
+			"t":   "$_id.t",
+		},
+	}, {
+		"$out": "newstatcounters",
+	}}).AllowDiskUse().All(&r)
+	if err != nil {
+		return errgo.Notef(err, "cannot squash stats")
+	}
+	if err := renameCollection(db, "newstatcounters", statCountersColl); err != nil {
+		return errgo.Mask(err)
+	}
+	return nil
+}
+
+func reorderStatsKeys(db *mgo.Database) error {
+	const maxQueue = 250
+
+	keepKeys := make(map[string]bool)
+	for _, k := range keepKinds {
+		key, err := statsKey(db, k)
+		if err != nil {
+			if errgo.Cause(err) == params.ErrNotFound {
```

Comment:
mgo.ErrNotFound

This implies this path hasn't been tested, please consider whether it needs to be.

Reply:
removed the case. it should never happen.